lib/utils/src/lib.rs

```diff
@@ -126,3 +146,32 @@ pub fn calc_capacity_sum(lock_hash: &[u8; 32], source: Source) -> u128 {
         .map(|cell| cell.capacity().unpack() as u128)
         .sum()
 }
+
+pub fn check_spore_address(
+    group_source: Source,
+    spore_address: action::Address,
+) -> Result<(), Error> {
+    let address = load_cell_lock(0, group_source)?;
+    let action::AddressUnion::Script(expected_script) = spore_address.to_enum();
+    if address.as_slice() != expected_script.as_slice() {
+        return Err(Error::SporeActionFieldMismatch);
+    }
+    Ok(())
+}
+
+pub fn extract_spore_action() -> Result<action::SporeAction, Error> {
+    let message = ckb_transaction_cobuild::fetch_message()
+        .map_err(|_| Error::InvliadCoBuildWitnessLayout)?
+        .ok_or(Error::InvliadCoBuildWitnessLayout)?;
+    let script_hash = load_script_hash()?;
+
+    let mut iter = message
+        .actions()
+        .into_iter()
+        .filter(|value| value.script_hash().as_slice() == script_hash.as_slice());
+    match (iter.next(), iter.next()) {
+        (Some(action), None) => action::SporeAction::from_slice(&action.data().raw_data())
+            .map_err(|_| Error::InvliadCoBuildMessage),
+        _ => Err(Error::InvliadCoBuildMessage),
```

Comment:
Error::SporeActionDuplicated is defined but not used.
Is it supposed to be used here?

Reply:
oh yes, after merging jjy's suggestion, it's changed